relay/pkg/block/forwarder.go

```diff
@@ -0,0 +1,143 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/tbtc/relay/pkg/btc"
+	"github.com/keep-network/tbtc/relay/pkg/chain"
+)
+
+// TODO: Make the following refactoring:
+//  - rename `block` package to `header` (align all logging and stuff)
+//  - rename `Forwarder` to `Relay`
+//  - rename `RunForwarder` to `StartRelay`
+
+const (
+	// Size of the headers queue.
+	headersQueueSize = 50
+
+	// Maximum size of processed headers batch.
+	headersBatchSize = 5
+
+	// Maximum time for which the pulling process will wait for a single header
+	// to be delivered by the headers queue.
+	headerTimeout = 1 * time.Second
+
+	// Block duration of a Bitcoin difficulty epoch.
+	difficultyEpochDuration = 2016
+
+	// Duration for which the forwarder should rest after performing
+	// a push action.
+	forwarderPushingSleepTime = 45 * time.Second
+)
+
+var logger = log.Logger("relay-block-forwarder")
+
+// Forwarder takes blocks from the Bitcoin chain and forwards them to the
+// given host chain.
+type Forwarder struct {
+	btcChain  btc.Handle
+	hostChain chain.Handle
+
+	processedHeaders int
+
+	headersQueue chan *btc.Header
+	errChan      chan error
+
+	loopExitHandler func()
+}
+
+// RunForwarder creates an instance of the block forwarder and runs its
+// processing loops. The lifecycle of the forwarder can be managed using the
+// passed context.
+func RunForwarder(
+	ctx context.Context,
+	btcChain btc.Handle,
+	hostChain chain.Handle,
+) *Forwarder {
+	loopCtx, cancelLoopCtx := context.WithCancel(ctx)
```

Comment:
Nope, `loopExitHandler` will be also called from the pulling loop (see #769). It's a mechanism that allows to completely shut down the entire forwarder in case one of the loops (pulling or pushing) raises an error.

Reply:
In this case, what do you think about passing cancel function as a parameter? It is cleaner approach IMO: we pass the context and a function to cancel it, just like it is returned by `context.WithCancel`:

```
pushingLoop(ctx context.Context, cancelFn context.CancelFn) 
```

Reply:
I wanted to avoid this because I think the loop itself should not decide about cancelling its context. In my opinion, it's the responsibility of the forwarder/relay itself as it controls the lifecycle of pulling/pushing loops. What the loop should do is to inform the forwarder about an abnormal exit and the forwarder should decide what to do - in this case, cancel the `loopCtx` but in other situation, it may want to restart the loop. Also, having `loopExitHandler` we can easily add other behavior which should take place on loop exit if desired.

Reply:
So maybe:
```
	go func() {
		forwarder.pushingLoop(loopCtx)
		cancelLoopCtx() // loop exited, cancel the contet
	}
```
?

I don't like we need to trust `pushingLoop` to call `loopExitHandler`. I do agree that it should not cancel the context, so the approach when the context is cancelled straight from `RunForwarder` in case one of the loops (currently we have just one loop, soon we'll have two) exits is IMO better - we do not need to trust those functions, everything is controlled by `RunForwarder` which triggered them.

Reply:
Yep, that's a much better one. Please see: https://github.com/keep-network/tbtc/commit/37c633f2384860b280f68b77636116d214e09584